File: macos/SureWord/Notes/NotesStore.swift
import Foundation

/// Shared, persisted cache for the notes library — a port of
/// `mobile/src/features/notes/notesStore.ts`.
///
/// Both the list and the editor read and write through this one object, so:
///
/// - the list renders the last snapshot instantly and revalidates silently;
/// - opening a note whose body has been seen before shows no spinner;
/// - an edit in the editor is visible in the list immediately.
///
/// List fetches return summary rows with no body. Bodies are merged in from
/// single-note fetches and saves, and `hasBody` marks the entries whose body
/// fields are real rather than summary placeholders.
@MainActor
@Observable
final class NotesStore {
    /// The app-wide instance. A single cache is the point: two would let the
    /// list and the editor disagree about the same note.
    static let shared = NotesStore(cacheURL: NotesStore.defaultCacheURL)

    private(set) var notes: [Note] = []
    private(set) var folders: [Folder] = []
    private(set) var tags: [Tag] = []
    /// True once the persisted cache has been read, or found absent.
    private(set) var isHydrated = false

    @ObservationIgnored private let cacheURL: URL?
    @ObservationIgnored private var hydrateTask: Task<Void, Never>?

    /// `cacheURL: nil` keeps the store entirely in memory, which is what the
    /// unit tests use.
    init(cacheURL: URL?) {
        self.cacheURL = cacheURL
    }

    // MARK: - Persistence

    private struct Payload: Codable {
        var notes: [Note]
        var folders: [Folder]
        var tags: [Tag]
    }

    static var defaultCacheURL: URL? {
        guard
            let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
        else { return nil }
        let directory = base.appendingPathComponent("SureWord", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory.appendingPathComponent("notes-cache.v1.json")
    }

    /// Read the persisted cache exactly once per app run.
    func hydrate() async {
        if let hydrateTask {
            await hydrateTask.value
            return
        }
        let task = Task { @MainActor in
            defer { isHydrated = true }
            guard let cacheURL, let data = try? Data(contentsOf: cacheURL) else { return }
            guard let payload = try? JSONDecoder().decode(Payload.self, from: data) else {
                // A corrupt cache must not block a network load.
                return
            }
            notes = payload.notes
            folders = payload.folders
            tags = payload.tags
        }
        hydrateTask = task
        await task.value
    }

    private func persist() {
        guard let cacheURL else { return }
        let payload = Payload(notes: notes, folders: folders, tags: tags)
        guard let data = try? JSONEncoder().encode(payload) else { return }
        // A full or unwritable disk must never break the UI, so the failure is
        // swallowed rather than surfaced — the cache is an optimisation.
        try? data.write(to: cacheURL, options: .atomic)
    }

    // MARK: - Mutations

    /// Replace the cache with a server snapshot. List rows are summaries, so a
    /// cached body is carried over while the server row is unchanged
    /// (`updatedAt` bumps on every edit); a changed row drops the stale body so
    /// the editor refetches it.
    func applyServerSnapshot(notes serverNotes: [Note], folders: [Folder], tags: [Tag]) {
        let previous = Dictionary(notes.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
        self.notes = serverNotes.map { note in
            guard
                let prior = previous[note.id],
                prior.hasBody,
                !note.hasBody,
                prior.updatedAt == note.updatedAt
            else { return note }
            var merged = note
            merged.content = prior.content
            merged.htmlContent = prior.htmlContent
            merged.hasBody = true
            return merged
        }
        self.folders = folders
        self.tags = tags
        isHydrated = true
        persist()
    }

    /// Refresh folders and tags without touching the cached notes.
    func applyFoldersAndTags(folders: [Folder], tags: [Tag]) {
        self.folders = folders
        self.tags = tags
        persist()
    }

    func cachedNote(id: String) -> Note? {
        notes.first { $0.id == id }
    }

    func upsert(_ note: Note) {
        if let index = notes.firstIndex(where: { $0.id == note.id }) {
            notes[index] = note
        } else {
            notes.insert(note, at: 0)
        }
        persist()
    }

    func patch(id: String, _ change: (inout Note) -> Void) {
        guard let index = notes.firstIndex(where: { $0.id == id }) else { return }
        change(&notes[index])
        persist()
    }

    func removeNote(id: String) {
        notes.removeAll { $0.id == id }
        persist()
    }

    func addFolder(_ folder: Folder) {
        folders.append(folder)
        persist()
    }

    func renameFolder(_ folder: Folder) {
        if let index = folders.firstIndex(where: { $0.id == folder.id }) {
            folders[index] = folder
        }
        persist()
    }

    /// Deleting a folder unfiles its notes, matching what
    /// `DELETE /api/folders/{id}` does on the server.
    func removeFolder(id: String) {
        folders.removeAll { $0.id == id }
        for index in notes.indices where notes[index].folderId == id {
            notes[index].folderId = nil
        }
        persist()
    }

    func addTag(_ tag: Tag) {
        tags.append(tag)
        persist()
    }

    /// The join rows cascade on the server, so the local copies go too.
    func removeTag(id: String) {
        tags.removeAll { $0.id == id }
        for index in notes.indices where notes[index].tagIds.contains(id) {
            notes[index].tagIds.removeAll { $0 == id }
        }
        persist()
    }

    /// Test seam: drop everything without touching disk semantics.
    func reset() {
        notes = []
        folders = []
        tags = []
    }
}

File: mobile/src/features/notes/notesStore.ts
import AsyncStorage from "@react-native-async-storage/async-storage";
import { useSyncExternalStore } from "react";
import type { Folder, Note, Tag } from "./types";

/**
 * Shared, persisted cache for the notes library. Both the list screen and the
 * note editor read from and write through to this module-level store, so:
 *
 * - the list renders the last snapshot instantly (hydrated from AsyncStorage)
 *   and revalidates silently in the background;
 * - the editor opens a previously-seen note without a spinner;
 * - edits made in the editor are visible in the list immediately, no
 *   pull-to-refresh needed.
 *
 * List fetches return summary rows (no content/htmlContent). Note bodies are
 * merged in from single-note fetches and saves; `hasBody` marks entries whose
 * body fields are real rather than summary placeholders.
 */

// v2: v1 rows predate `aliases`/`properties`, and reading one back would hand
// the UI a Note whose array fields are undefined. Dropping the old cache costs
// one silent revalidation.
const STORAGE_KEY = "sureword.notes-cache.v2";

export interface NotesSnapshot {
	notes: Note[];
	folders: Folder[];
	tags: Tag[];
	/** True once the persisted cache has been read (or found absent). */
	hydrated: boolean;
}

let snapshot: NotesSnapshot = { notes: [], folders: [], tags: [], hydrated: false };

const listeners = new Set<() => void>();

function emit() {
	for (const listener of listeners) listener();
}

function persist() {
	AsyncStorage.setItem(
		STORAGE_KEY,
		JSON.stringify({ notes: snapshot.notes, folders: snapshot.folders, tags: snapshot.tags })
	).catch(() => {
		// A full/unavailable store must never break the UI.
	});
}

function setSnapshot(next: Omit<NotesSnapshot, "hydrated">) {
	snapshot = { ...next, hydrated: true };
	persist();
	emit();
}

let hydratePromise: Promise<void> | null = null;

/**
 * Bumped by every clear. A read issued before a clear is still in flight after
 * it, and would otherwise resolve straight into setSnapshot and hand the new
 * account the previous account's notes.
 */
let generation = 0;

/** Read the persisted cache exactly once per app run. */
export function hydrateNotesCache(): Promise<void> {
	if (!hydratePromise) {
		const startedAt = generation;
		hydratePromise = (async () => {
			try {
				const raw = await AsyncStorage.getItem(STORAGE_KEY);
				if (generation !== startedAt) return;
				if (raw) {
					const parsed = JSON.parse(raw) as Partial<NotesSnapshot>;
					setSnapshot({
						notes: parsed.notes ?? [],
						folders: parsed.folders ?? [],
						tags: parsed.tags ?? [],
					});
				}
			} catch {
				// Corrupt or unreadable cache: fall through to a network load.
			} finally {
				if (!snapshot.hydrated) {
					snapshot = { ...snapshot, hydrated: true };
				}
				emit();
			}
		})();
	}
	return hydratePromise;
}

/**
 * Drop every cached row and the persisted blob behind it. Called when the
 * signed-in account is not the one this cache was written for, and on sign-out:
 * notes are private study, and the next account must never see them.
 *
 * `hydratePromise` is replaced rather than cleared so a screen mounting later
 * cannot re-read the file we are deleting; the snapshot is already marked
 * hydrated, so the library falls straight through to a server load.
 */
export function clearNotesCache(): Promise<void> {
	generation += 1;
	snapshot = { notes: [], folders: [], tags: [], hydrated: true };
	hydratePromise = Promise.resolve();
	emit();
	return AsyncStorage.removeItem(STORAGE_KEY).catch(() => {
		// An unwritable store must never break sign-out; the snapshot is
		// already empty, so nothing is shown from it either way.
	});
}

/**
 * Replace the cache with a server snapshot. List rows are summaries, so a
 * cached body is carried over while the server row is unchanged (updatedAt
 * bumps on every edit); a changed row drops the stale body so the editor
 * refetches it.
 */
export function applyServerSnapshot(notes: Note[], folders: Folder[], tags: Tag[]) {
	const prevById = new Map(snapshot.notes.map((note) => [note.id, note]));
	const merged = notes.map((note) => {
		const prev = prevById.get(note.id);
		if (
			prev &&
			prev.hasBody &&
			!note.hasBody &&
			prev.updatedAt === note.updatedAt
		) {
			return { ...note, content: prev.content, htmlContent: prev.htmlContent, hasBody: true };
		}
		return note;
	});
	setSnapshot({ notes: merged, folders, tags });
}

export function getCachedNote(id: string): Note | null {
	return snapshot.notes.find((note) => note.id === id) ?? null;
}

/** Refresh folders/tags without touching the cached notes. */
export function applyFoldersAndTags(folders: Folder[], tags: Tag[]) {
	setSnapshot({ ...snapshot, folders, tags });
}

export function upsertNoteInCache(note: Note) {
	const exists = snapshot.notes.some((entry) => entry.id === note.id);
	setSnapshot({
		...snapshot,
		notes: exists
			? snapshot.notes.map((entry) => (entry.id === note.id ? note : entry))
			: [note, ...snapshot.notes],
	});
}

export function patchNoteInCache(id: string, changes: Partial<Note>) {
	setSnapshot({
		...snapshot,
		notes: snapshot.notes.map((note) => (note.id === id ? { ...note, ...changes } : note)),
	});
}

export function removeNoteFromCache(id: string) {
	setSnapshot({ ...snapshot, notes: snapshot.notes.filter((note) => note.id !== id) });
}

export function addFolderToCache(folder: Folder) {
	setSnapshot({ ...snapshot, folders: [...snapshot.folders, folder] });
}

export function removeFolderFromCache(id: string) {
	setSnapshot({
		...snapshot,
		folders: snapshot.folders.filter((folder) => folder.id !== id),
		notes: snapshot.notes.map((note) =>
			note.folderId === id ? { ...note, folderId: null } : note
		),
	});
}

export function addTagToCache(tag: Tag) {
	setSnapshot({ ...snapshot, tags: [...snapshot.tags, tag] });
}

export function removeTagFromCache(id: string) {
	setSnapshot({
		...snapshot,
		tags: snapshot.tags.filter((tag) => tag.id !== id),
		notes: snapshot.notes.map((note) =>
			note.tagIds.includes(id)
				? { ...note, tagIds: note.tagIds.filter((tagId) => tagId !== id) }
				: note
		),
	});
}

function subscribe(listener: () => void): () => void {
	listeners.add(listener);
	return () => {
		listeners.delete(listener);
	};
}

export function useNotesSnapshot(): NotesSnapshot {
	return useSyncExternalStore(subscribe, () => snapshot);
}
